### PR TITLE
Add the test cases for Restricted Security Mode

### PR DESCRIPTION
**Pull request by taoliult**
This PR is for adding the test cases for Restricted Security Mode, to cover the restricted security mode providers and properties related test scenarios.

### CODE REVIEW

**Comment by llxia** [comment]
Just for the record, the tests can only run locally atm. We will have a separate PR to enable the tests in our test pipeline.

**Comment by taoliult** [comment]
@keithc-ca Please help to review and advise.

File: closed/test/jdk/TEST.ROOT [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * ===========================================================================
+ */
+
+# Path to libraries in the topmost test directory. This is needed so @library
+# does not need ../../../ notation to reach them
```
Comment (keithc-ca) [inline closed/test/jdk/TEST.ROOT:26]: Please finish each sentence with a period.

File: closed/test/jdk/openj9/internal/security/TestProperties.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+public class TestProperties {
+
+    private static Stream<Arguments> patternMatches_expectedExitValue1() {
+        return Stream.of(
+                // 1 - Test profile - base profile misspell properties
+                Arguments.of("Test-Profile.Base",
+                        System.getProperty("test.src") + "/property-java.security",
+                        "The property names: RestrictedSecurity.Test-Profile.Base.tls.disabledAlgorithmsWrongTypo " +
+                                "in profile RestrictedSecurity.Test-Profile.Base \\(or a base profile\\) are not recognized"),
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProperties.java:52]: Please put `+` at the beginning of continuations:
```java
                        "The property names: RestrictedSecurity.Test-Profile.Base.tls.disabledAlgorithmsWrongTypo "
                            + "in profile RestrictedSecurity.Test-Profile.Base \\(or a base profile\\) are not recognized"),
```

File: closed/test/jdk/openj9/internal/security/TestProperties.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("patternMatches_expectedExitValue1")
+    public void shouldContain_expectedExitValue1(String customprofile, String securityPropertyFile, String expected) throws Exception {
+        OutputAnalyzer outputAnalyzer = ProcessTools.executeTestJava(
+                "-Dsemeru.fips=true",
+                "-Dsemeru.customprofile=" + customprofile,
+                "-Djava.security.properties=" + securityPropertyFile,
+                //"-Djava.security.debug=semerufips",
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProperties.java:165]: I don't think this comment should be here.

File: closed/test/jdk/openj9/internal/security/TestProperties.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                "TestProperties"
+        );
+        outputAnalyzer.reportDiagnosticSummary();
+        outputAnalyzer.shouldHaveExitValue(1).shouldMatch(expected);
+    }
+
+    public static void main(String[] args) throws Exception {
+        // Something to trigger "properties" debug output
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProperties.java:173]: Missing period.

File: closed/test/jdk/openj9/internal/security/TestProperties.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        try {
+            Provider p[] = Security.getProviders();
+            for (int i = 0; i < p.length; i++) {
+                System.out.println("Provider Name: " + p[i].getName());
+                System.out.println("Provider Version: " + p[i].getVersion());
+            }
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProperties.java:179]: Please use enhanced for loop:
```java
            for (Provider provider : Security.getProviders()) {
                System.out.println("Provider Name: " + provider.getName());
                System.out.println("Provider Version: " + provider.getVersion());
            }
```

File: closed/test/jdk/openj9/internal/security/TestProperties.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+                "-Dsemeru.customprofile=" + customprofile,
+                "-Djava.security.properties=" + securityPropertyFile,
+                //"-Djava.security.debug=semerufips",
+                "TestProperties"
+        );
+        outputAnalyzer.reportDiagnosticSummary();
+        outputAnalyzer.shouldHaveExitValue(1).shouldMatch(expected);
+    }
+
+    public static void main(String[] args) throws Exception {
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProperties.java:172]: This doesn't throw `Exception`.

File: closed/test/jdk/openj9/internal/security/property-java.security [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * ===========================================================================
+ */
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/property-java.security:23]: I don't expect this comment is legal in a properties file: use `#` like below.

**Review by keithc-ca — CHANGES_REQUESTED** [review]
Why are these tests here instead of in OpenJ9?

File: closed/test/jdk/openj9/internal/security/property-java.security (hunk 7 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/property-java.security:23]: @keithc-ca 
Do we need to add the copyright info in this test case used java.security file? I checked the java.security file under the java.base which we added those restricted security profiles, there is no copyright info.

File: closed/test/jdk/TEST.ROOT (hunk 1 above)
Comment (taoliult) [reply closed/test/jdk/TEST.ROOT]: Done.

File: closed/test/jdk/openj9/internal/security/TestProperties.java (hunk 2 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProperties.java]: Done.

File: closed/test/jdk/openj9/internal/security/TestProperties.java (hunk 4 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProperties.java]: Done.

File: closed/test/jdk/openj9/internal/security/TestProperties.java (hunk 5 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProperties.java]: Done.

File: closed/test/jdk/openj9/internal/security/TestProperties.java (hunk 6 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProperties.java]: Done.

File: closed/test/jdk/openj9/internal/security/property-java.security (hunk 7 above)
Comment (keithc-ca) [reply closed/test/jdk/openj9/internal/security/property-java.security:23]: Yes, I think this file should have a copyright notice; it just needs to use the right comment syntax.

File: closed/test/jdk/TEST.ROOT [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * ===========================================================================
+ */
```
Comment (keithc-ca) [inline closed/test/jdk/TEST.ROOT:23]: Comments begin with `#` in files such as this.

File: closed/test/jdk/openj9/internal/security/TestProperties.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * ===========================================================================
+ */
+
+ /*
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProperties.java:25]: No space before `/*`, please.

File: closed/test/jdk/openj9/internal/security/TestProperties.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+import java.util.stream.Stream;
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+public class TestProperties {
+
+    private static Stream<Arguments> patternMatches_expectedExitValue1() {
+        return Stream.of(
+                // 1 - Test profile - base profile misspell properties
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProperties.java:48]: Please finish each sentence with a period.

File: closed/test/jdk/openj9/internal/security/TestProperties.java [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+                "TestProperties"
+        );
+        outputAnalyzer.reportDiagnosticSummary();
+        outputAnalyzer.shouldHaveExitValue(1).shouldMatch(expected);
+    }
+
+    public static void main(String[] args) {
+        // Something to trigger "properties" debug output.
+        try {
+            Provider p[] = Security.getProviders();
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProperties.java:174]: Please remove this unused local.

File: closed/test/jdk/openj9/internal/security/TestProviders.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * ===========================================================================
+ */
+
+ /*
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProviders.java:25]: No space before `/*`, please.

File: closed/test/jdk/openj9/internal/security/TestProviders.java [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+import java.util.stream.Stream;
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+public class TestProviders {
+
+    private static Stream<Arguments> patternMatches_expectedExitValue0() {
+        return Stream.of(
+                // Test strict profile provider list
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProviders.java:48]: Please finish each sentence with a period.

File: closed/test/jdk/openj9/internal/security/TestProviders.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+                "-Djava.security.properties=" + securityPropertyFile,
+                "TestProviders"
+        );
+        outputAnalyzer.reportDiagnosticSummary();
+        outputAnalyzer.shouldHaveExitValue(1).shouldMatch(expected);
+    }
+
+    public static void main(String[] args) {
+        try {
+            Provider p[] = Security.getProviders();
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProviders.java:151]: Please remove this unused local.

File: closed/test/jdk/openj9/internal/security/provider-java.security [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * ===========================================================================
+ */
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/provider-java.security:23]: Comments in this file begin with `#`.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by keithc-ca** [comment]
> Why are these tests here instead of in OpenJ9?

**Comment by pshipton** [comment]
The code is here, why shouldn't the tests be here? It's a class library thing, not a VM thing. In theory it could work with whatever VM you happen to be using.

**Comment by keithc-ca** [comment]
Ok, but that just means we'll have many (i.e. six) copies of this code; doesn't seem like a good plan to me.

File: closed/test/jdk/openj9/internal/security/property-java.security (hunk 7 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/property-java.security]: Done.

File: closed/test/jdk/TEST.ROOT (hunk 8 above)
Comment (taoliult) [reply closed/test/jdk/TEST.ROOT]: Done.

File: closed/test/jdk/openj9/internal/security/TestProperties.java (hunk 9 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProperties.java:24]: Done.

File: closed/test/jdk/openj9/internal/security/TestProperties.java (hunk 10 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProperties.java]: Done.

File: closed/test/jdk/openj9/internal/security/TestProperties.java (hunk 11 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProperties.java]: Done.

File: closed/test/jdk/openj9/internal/security/TestProviders.java (hunk 12 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProviders.java:24]: Done.

File: closed/test/jdk/openj9/internal/security/TestProviders.java (hunk 13 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProviders.java]: Done.

File: closed/test/jdk/openj9/internal/security/TestProviders.java (hunk 14 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProviders.java]: Done.

File: closed/test/jdk/openj9/internal/security/provider-java.security (hunk 15 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/provider-java.security]: Done.

**Comment by taoliult** [comment]
> Why are these tests here instead of in OpenJ9?

When I started working on these test cases, I reached out to Lan and Peter, and they suggested the` /closed/test/ `directory should be the place for these tests codes. 

@llxia If there is any structure changes needed when you are adding the test target to OpenJDK playlist, just let me know.

**Comment by llxia** [comment]
I think the test code should be located near the feature code for better organization. Additionally, the openj9 repo currently contains only functional tests. Including OpenJDK tests in the openj9 repo would require supporting the setup of an OpenJDK test environment within it, which is not ideal.

I agree that having multiple copies of this code is not ideal; however, that's how extension repos are structured.

File: closed/test/jdk/TEST.ROOT [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+# version 2 for more details (a copy is included in the LICENSE file that
+# accompanied this code).
+#
+# You should have received a copy of the GNU General Public License version
+# 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+#
+# ===========================================================================
+#
```
Comment (keithc-ca) [inline closed/test/jdk/TEST.ROOT:23]: Please remove lines 1, 5, 21 and 23 for consistency with other copyright notices.

File: closed/test/jdk/openj9/internal/security/property-java.security [hunk 17]
```diff
@@ -0,0 +1,531 @@
...
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+# version 2 for more details (a copy is included in the LICENSE file that
+# accompanied this code).
+#
+# You should have received a copy of the GNU General Public License version
+# 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+#
+# ===========================================================================
+#
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/property-java.security:23]: Please remove lines 1, 5, 21 and 23 for consistency with other copyright notices.

File: closed/test/jdk/openj9/internal/security/provider-java.security [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+# version 2 for more details (a copy is included in the LICENSE file that
+# accompanied this code).
+#
+# You should have received a copy of the GNU General Public License version
+# 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+#
+# ===========================================================================
+#
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/provider-java.security:23]: Please remove lines 1, 5, 21 and 23 for consistency with other copyright notices.

File: closed/test/jdk/openj9/internal/security/TestProperties.java (hunk 9 above)
Comment (keithc-ca) [reply closed/test/jdk/openj9/internal/security/TestProperties.java:24]: The blank line was fine, but the first line of the second comment should not start with a space.

File: closed/test/jdk/openj9/internal/security/TestProperties.java [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+                        "Test-Profile-NotExist.Base is not present in the java.security file."),
+                // 5 - Test profile - Multi Default profile.
+                Arguments.of("Test-Profile-MultiDefault",
+                        System.getProperty("test.src") + "/property-java.security",
+                        "Multiple default RestrictedSecurity profiles for Test-Profile-MultiDefault"),
+                // 6 - Test profile - no default profile.
+                Arguments.of("Test-Profile-NoDefault",
+                        System.getProperty("test.src") + "/property-java.security",
+                        "No default RestrictedSecurity profile was found for Test-Profile-NoDefault"),
+                // 7 - Test profile - base profile not exist.
```
Comment (keithc-ca) [inline closed/test/jdk/openj9/internal/security/TestProperties.java:75]: Missing "does" in "base profile does not exist".

File: closed/test/jdk/openj9/internal/security/TestProviders.java (hunk 12 above)
Comment (keithc-ca) [reply closed/test/jdk/openj9/internal/security/TestProviders.java:24]: The blank line was fine, but the first line of the second comment should not start with a space.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: closed/test/jdk/openj9/internal/security/property-java.security (hunk 17 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/property-java.security]: Done.

File: closed/test/jdk/openj9/internal/security/provider-java.security (hunk 18 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/provider-java.security]: Done.

File: closed/test/jdk/openj9/internal/security/TestProperties.java (hunk 19 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProperties.java]: Done.

File: closed/test/jdk/openj9/internal/security/TestProviders.java (hunk 12 above)
Comment (taoliult) [reply closed/test/jdk/openj9/internal/security/TestProviders.java]: Done.

**Review by keithc-ca — APPROVED** [review]

**Comment by taoliult** [comment]
Working on the JDK8 PR.

**Pull request merged**